core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -58,11 +68,211 @@ case class PSBT(
+          inputMaps(i)
+            .getRecords[InputPSBTRecord.WitnessScript](WitnessScriptKeyId)
+            .size == 1
+
+        if (outIsWitnessScript || hasWitScript) {
+          elements = elements.filterNot(_.key.startsWith(hex"01")) :+ WitnessUTXO(
+            out)
+        } else {
+          elements = elements.filterNot(_.key.startsWith(hex"00")) :+ NonWitnessOrUnknownUTXO(
+            tx)
+        }
+      }
+
+      val newInputMaps =
+        inputMaps.patch(i, Seq(InputPSBTMap(elements)), 1)
+      PSBT(globalMap, newInputMaps, outputMaps)
+    }
+  }
+
+  /**
+    * @param script ScriptPubKey to add to PSBT
+    * @param i index of inputMap to add script to
+    * @return PSBT with added script
+    */
+  def addScriptToInput(script: ScriptPubKey, i: Int): PSBT = {
+    require(
+      i < inputMaps.size,
+      s"i must be less than the number of input maps present in the psbt, $i >= ${inputMaps.size}")
+    if (inputMaps(i).isFinalized) {
+      this
+    } else {
+      var elements = inputMaps(i).elements
+
+      val isWitScript = WitnessScriptPubKey.isWitnessScriptPubKey(script.asm)
+      val redeemScript = inputMaps(i).getRecords[InputPSBTRecord.RedeemScript](
+        PSBTInputKeyId.RedeemScriptKeyId)
+      val hasWitScript = redeemScript.size == 1 && WitnessScriptPubKey
+        .isWitnessScriptPubKey(redeemScript.head.redeemScript.asm)
+
+      if (!isWitScript && hasWitScript) {
+        elements = elements.filterNot(_.key.startsWith(hex"05")) :+ InputPSBTRecord
+          .WitnessScript(script)
+      } else {
+        elements = elements.filterNot(_.key.startsWith(hex"04")) :+ InputPSBTRecord
+          .RedeemScript(script)
+      }
+
+      val newMap = InputPSBTMap(elements).slimMap(transaction.inputs(i))
+      val newInputMaps = inputMaps.patch(i, Seq(newMap), 1)
+
+      PSBT(globalMap, newInputMaps, outputMaps)
+    }
+  }
+
+  /**
+    * @param script ScriptPubKey to add to PSBT
+    * @param i index of outputMap to add script to
+    * @return PSBT with added script
+    */
+  def addScriptToOutput(script: ScriptPubKey, i: Int): PSBT = {
+    require(
+      i < outputMaps.size,
+      s"i must be less than the number of output maps present in the psbt, $i >= ${outputMaps.size}")
+    if (isFinalized) {
+      this
+    } else {
+      var elements = outputMaps(i).elements
+
+      val isWitScript = WitnessScriptPubKey.isWitnessScriptPubKey(script.asm)
+      val redeemScript =
+        outputMaps(i).getRecords[psbt.OutputPSBTRecord.RedeemScript](
+          PSBTOutputKeyId.RedeemScriptKeyId)
+      val hasWitScript = redeemScript.size == 1 && WitnessScriptPubKey
+        .isWitnessScriptPubKey(redeemScript.head.redeemScript.asm)
+
+      if (!isWitScript && hasWitScript) {
+        elements = elements.filterNot(_.key.startsWith(hex"01")) :+ OutputPSBTRecord
+          .WitnessScript(script)
+      } else {
+        elements = elements.filterNot(_.key.startsWith(hex"00")) :+ OutputPSBTRecord
```

Comment:
I think for all of these `key.startsWith` wit would be nice to have some simple `ADT` someplace wrapping these bytes so that we could give them nice readable names rather than using numbers